File: npm/rules/ci4/package_knowledge/render.mjs
/**
 * Рендерить deterministic Markdown і manifest-проєкції package knowledge graph.
 *
 * Модуль не аналізує source, не викликає LLM і не публікує файли. Він створює
 * повний candidate file map, а publication лишається відповідальністю
 * `publish.mjs` після окремої validation-перевірки.
 */

import { createHash } from 'node:crypto'

import { createImpactSlice } from './impact.mjs'
import { serializeKnowledgeGraph } from './normalized-graph.mjs'
import { collectReachableNodeIds } from './topic-discovery.mjs'
import { applyAutogenUpdates, parseKnowledgeZones, zoneHash } from './zones.mjs'

const MANIFEST_PATH = 'docs/.docgen/manifest.json'
const PAGE_KIND_PATHS = {
  capability: 'docs/explanation/capabilities',
  contract: 'docs/reference/contracts',
  process: 'docs/explanation/processes'
}
const PAGE_KIND_LABELS = {
  capability: 'Можливість',
  contract: 'Контракт',
  process: 'Процес'
}
const CLAIM_SECTION_TITLES = Object.freeze({
  purpose: 'Призначення',
  actor: 'Actors',
  trigger: 'Trigger',
  precondition: 'Передумови',
  step: 'Основний потік',
  'business-rule': 'Business rules',
  'state-change': 'Зміни стану',
  integration: 'Integration boundaries',
  outcome: 'Результати',
  'alternative-flow': 'Alternative flows',
  'error-flow': 'Error flows',
  responsibility: 'Відповідальності',
  config: 'Configuration',
  persistence: 'Persistence'
})

/**
 * Створює stable renderer diagnostic.
 * @param {string} code machine-readable code
 * @param {string} detail explanation
 * @param {string | null} [path] affected candidate path
 * @returns {{code: string, detail: string, path: string | null}} diagnostic
 */
function diagnostic(code, detail, path = null) {
  return { code, detail, path }
}

/**
 * Повертає короткий filesystem-safe stable topic token.
 * @param {string} topicId canonical topic ID
 * @returns {string} path-safe deterministic token
 */
function topicToken(topicId) {
  return createHash('sha256').update(topicId).digest('hex').slice(0, 24)
}

/**
 * Порівнює strings у стабільному порядку.
 * @param {string} left first value
 * @param {string} right second value
 * @returns {number} comparison result
 */
function compareStrings(left, right) {
  return left.localeCompare(right)
}

/**
 * Рекурсивно canonicalizes object keys for byte-stable manifest JSON.
 * @param {unknown} value serializable value
 * @returns {unknown} canonical value
 */
function canonicalize(value) {
  if (Array.isArray(value)) return value.map(item => canonicalize(item))
  if (!value || typeof value !== 'object') return value
  return Object.fromEntries(
    Object.entries(value)
      .toSorted(([left], [right]) => compareStrings(left, right))
      .map(([key, item]) => [key, canonicalize(item)])
  )
}

/**
 * Sorts graph collections without changing their schema or traceability data.
 * @param {Record<string, unknown>} graph source knowledge graph
 * @returns {Record<string, unknown>} manifest-ready graph
 */
function canonicalGraph(graph) {
  const output = canonicalize(graph)
  for (const key of ['nodes', 'edges', 'claims', 'topics', 'gaps', 'evidence']) {
    if (Array.isArray(output[key]))
      output[key] = output[key].toSorted((left, right) => compareStrings(left.id, right.id))
  }
  return output
}

/**
 * Reads only human-safe node titles. Private symbols remain manifest-only.
 * @param {Record<string, unknown>} graph knowledge graph
 * @returns {Set<string>} private names and IDs that may not occur in Markdown
 */
function privateNames(graph) {
  return new Set(
    (Array.isArray(graph.nodes) ? graph.nodes : [])
      .filter(node => node?.visibility === 'private')
      .flatMap(node => [node.name, node.id].filter(value => typeof value === 'string' && value !== ''))
  )
}

/**
 * Returns a display string only when it cannot leak a private symbol name.
 * @param {unknown} value candidate text
 * @param {Set<string>} hiddenNames private names
 * @param {string} fallback safe replacement
 * @returns {string} privacy-safe text
 */
function safeText(value, hiddenNames, fallback) {
  if (typeof value !== 'string' || value === '' || [...hiddenNames].some(name => value.includes(name))) return fallback
  return value
}

/**
 * Serializes an evidence-backed claim value without leaking private symbols.
 * @param {unknown} value structured claim value
 * @param {Set<string>} hiddenNames private names
 * @returns {string} safe compact value
 */
function safeValue(value, hiddenNames) {
  const rendered = typeof value === 'string' ? value : JSON.stringify(canonicalize(value))
  return safeText(rendered, hiddenNames, 'підтверджене значення')
}

/**
 * Wraps generated content in an AUTOGEN zone with a stable hash.
 * @param {string} id zone ID
 * @param {string} content generated Markdown
 * @returns {string} complete Markdown zone
 */
function autogenZone(id, content) {
  return `<!-- AUTOGEN:start id="${id}" hash="${zoneHash(content)}" -->${content}<!-- AUTOGEN:end id="${id}" -->`
}

/**
 * Updates an existing explicitly zoned page or creates a new generated page.
 * @param {{path: string, zoneId: string, content: string, existing: string | undefined}} input page projection
 * @returns {{ok: true, markdown: string} | {ok: false, diagnostics: Array<Record<string, unknown>>}} rendered page
 */
function renderPage({ path, zoneId, content, existing }) {
  if (existing === undefined) return { ok: true, markdown: autogenZone(zoneId, content) }
  if (typeof existing !== 'string') {
    return {
      ok: false,
      diagnostics: [diagnostic('invalid-existing-page', 'Existing page має бути Markdown string.', path)]
    }
  }
  const parsed = parseKnowledgeZones(existing, path)
  if (!parsed.ok) return parsed
  const zone = parsed.zones.find(item => item.kind === 'AUTOGEN' && item.id === zoneId)
  if (!zone) {
    return {
      ok: false,
      diagnostics: [diagnostic('autogen-zone-required', `Existing page має містити AUTOGEN ${zoneId}.`, path)]
    }
  }
  return applyAutogenUpdates(existing, { [zoneId]: content }, path)
}

/**
 * Returns graph topics that have a supported dedicated Markdown page.
 * @param {Record<string, unknown>} graph knowledge graph
 * @returns {Array<Record<string, unknown>>} sorted page topics
 */
function pageTopics(graph) {
  return (Array.isArray(graph.topics) ? graph.topics : [])
    .filter(topic => topic && Object.hasOwn(PAGE_KIND_PATHS, topic.kind) && typeof topic.id === 'string')
    .toSorted((left, right) => compareStrings(left.id, right.id))
}

/**
 * Groups a stable topic list by kind for index navigation.
 * @param {Array<Record<string, unknown>>} topics rendered topics
 * @returns {Map<string, Array<Record<string, unknown>>>} kind → topics
 */
function topicsByKind(topics) {
  const result = new Map()
  for (const topic of topics) {
    const items = result.get(topic.kind) ?? []
    items.push(topic)
    result.set(topic.kind, items)
  }
  return result
}

/**
 * Determines whether graph evidence supports a useful architecture projection.
 * @param {Record<string, unknown>} graph knowledge graph
 * @returns {boolean} whether architecture page is meaningful
 */
function needsArchitecture(graph) {
  const nodes = Array.isArray(graph.nodes) ? graph.nodes : []
  const responsibilities = nodes.filter(node => node?.domainId === graph.domain.id && node.kind === 'component')
  const boundaries = nodes.filter(
    node => node?.domainId === graph.domain.id && node.kind === 'integration' && node.visibility === 'external'
  )
  return responsibilities.length > 1 || boundaries.length > 0
}

/**
 * Renders a compact fact list or an explicit absence statement.
 * @param {string[]} facts evidence-backed facts
 * @param {string} fallback absence text
 * @returns {string} Markdown list or fallback
 */
function factList(facts, fallback) {
  return facts.length > 0 ? facts.map(fact => `- ${fact}`).join('\n') : fallback
}

/** Рендерить тільки присутні evidence-backed claim categories у stable order. */
function claimSections(claims, hiddenNames, fallback) {
  const byPredicate = new Map()
  for (const claim of claims) {
    const predicate = safeText(claim.predicate, hiddenNames, 'evidence-backed behavior')
    const values = byPredicate.get(predicate) ?? []
    values.push(safeValue(claim.value, hiddenNames))
    byPredicate.set(predicate, values)
  }
  const known = Object.keys(CLAIM_SECTION_TITLES).filter(predicate => byPredicate.has(predicate))
  const unknown = [...byPredicate.keys()].filter(predicate => !Object.hasOwn(CLAIM_SECTION_TITLES, predicate)).toSorted(compareStrings)
  const sections = [...known, ...unknown].map(predicate => {
    const title = CLAIM_SECTION_TITLES[predicate] ?? 'Інші підтверджені факти'
    const facts = [...new Set(byPredicate.get(predicate))]
      .map(value => (Object.hasOwn(CLAIM_SECTION_TITLES, predicate) ? value : `${predicate}: ${value}.`))
      .toSorted(compareStrings)
    return `## ${title}\n\n${factList(facts, fallback)}`
  })
  return sections.length > 0 ? sections.join('\n\n') : fallback
}

/* eslint-disable unicorn/no-array-callback-reference -- named claim formatter is intentionally reused for both layers */
/**
 * Collects topic-local public facts and reverse impact paths.
 * @param {{graph: Record<string, unknown>, topic: Record<string, unknown>, hiddenNames: Set<string>}} input render context
 * @returns {{implementedClaims: object[], expectedClaims: object[], outcomes: string[], contracts: string[], gaps: string[], paths: string[]}} safe topic facts
 */
function topicFacts({ graph, topic, hiddenNames }) {
  const reachable = new Set(collectReachableNodeIds(graph, topic.anchorIds))
  const claims = graph.claims.filter(claim => reachable.has(claim.subjectId))
  const publicNodes = graph.nodes.filter(node => reachable.has(node.id) && node.visibility !== 'private')
  const namesFor = kind =>
    publicNodes
      .filter(node => node.kind === kind)
      .map(node => safeText(node.name, hiddenNames, kind === 'outcome' ? 'Confirmed outcome' : 'External contract'))
      .toSorted(compareStrings)
  const localClaimIds = new Set(claims.map(claim => claim.id))
  const gaps = graph.gaps
    .filter(gap => localClaimIds.has(gap.expectedClaimId) || gap.implementedClaimIds?.some(id => localClaimIds.has(id)))
    .map(gap => `Status: ${gap.status}.`)
    .toSorted(compareStrings)
  const impact = createImpactSlice({ graph, topics: graph.topics, topicId: topic.id })
  const paths = impact.ok
    ? [...impact.slice.files, ...impact.slice.tests, ...impact.slice.configs].toSorted(compareStrings)
    : []
  return {
    implementedClaims: claims.filter(claim => claim.layer === 'implemented').toSorted((left, right) => compareStrings(left.id, right.id)),
    expectedClaims: claims.filter(claim => claim.layer === 'expected').toSorted((left, right) => compareStrings(left.id, right.id)),
    implemented: claims
      .filter(claim => claim.layer === 'implemented')
      .map(claim => `${safeText(claim.predicate, hiddenNames, 'evidence-backed behavior')}: ${safeValue(claim.value, hiddenNames)}.`)
      .toSorted(compareStrings),
    expected: claims
      .filter(claim => claim.layer === 'expected')
      .map(claim => `${safeText(claim.predicate, hiddenNames, 'evidence-backed behavior')}: ${safeValue(claim.value, hiddenNames)}.`)
      .toSorted(compareStrings),
    outcomes: namesFor('outcome'),
    contracts: namesFor('integration'),
    gaps,
    paths
  }
}
/* eslint-enable unicorn/no-array-callback-reference */

/**
 * Builds a self-contained AS-IS fragment for one discovered topic.
 * @param {{graph: Record<string, unknown>, topic: Record<string, unknown>, hiddenNames: Set<string>}} input render context
 * @returns {string} generated Markdown body
 */
function renderTopic({ graph, topic, hiddenNames }) {
  const label = PAGE_KIND_LABELS[topic.kind]
  const title = safeText(topic.title, hiddenNames, `${label} домену`)
  const facts = topicFacts({ graph, topic, hiddenNames })
  const aliases =
    Array.isArray(topic.aliases) && topic.aliases.length > 0
      ? `\n\nПопередні stable aliases: ${topic.aliases.length}.`
      : ''
  if (facts.implementedClaims.length > 0 || facts.expectedClaims.length > 0) {
    return `# ${label}: ${title}\n\n## Implemented AS-IS\n\n${claimSections(facts.implementedClaims, hiddenNames, 'Немає evidence-backed implemented behavioral claims для цього topic.')}\n\n## Outcomes і contracts\n\nOutcomes:\n${factList(facts.outcomes, 'Немає підтвердженого public outcome.')}\n\nContracts:\n${factList(facts.contracts, 'Немає підтвердженого external contract.')}\n\n## Affected paths\n\n${factList(
      facts.paths.map(path => `\`${path}\``),
      'Reverse impact paths відсутні у поточній graph projection.'
    )}\n\n## Expected behavior\n\n${claimSections(facts.expectedClaims, hiddenNames, 'Для topic немає explicit expected claim.')}\n\n## Local implementation gaps\n\n${factList(facts.gaps, 'Для topic немає actionable implementation gaps.')}${aliases}\n`
  }
  return `# ${label}: ${title}\n\n## Implemented AS-IS\n\nЦей self-contained fragment описує підтверджену поточну поведінку ${label.toLowerCase()} у domain \`${graph.domain.name}\`. Він не припускає intent поза evidence graph.\n\n## Призначення\n\n${title} надає evidence-backed boundary для зміни та перевірки поведінки domain.\n\n## Actors і trigger\n\nПотік починається з підтвердженого topic anchor і завершується зафіксованим результатом або external contract boundary.\n\n## Передумови\n\nВхід до ${label.toLowerCase()} доступний у межах owning domain, а потрібні integration boundaries представлені у traceability manifest.\n\n## Implemented facts\n\n${factList(facts.implemented, 'Для topic немає окремого implemented claim; AS-IS обмежений evidence-backed graph boundary.')}\n\n## Outcomes і contracts\n\nOutcomes:\n${factList(facts.outcomes, 'Немає окремо названого public outcome.')}\n\nContracts:\n${factList(facts.contracts, 'Немає external contract у reachable graph.')}\n\n## Affected paths\n\n${factList(
    facts.paths.map(path => `\`${path}\``),
    'Reverse impact paths відсутні у поточній graph projection.'
  )}\n\n## Alternative flows і rules\n\nAlternative та error-flow details відображаються лише тоді, коли їх представляють graph edges і claims; цей fragment не додає непідтверджених сценаріїв.\n\n## Expected behavior\n\n${factList(facts.expected, 'Для topic немає explicit expected claim. Відсутність expectation не створює implementation gap.')}\n\n## Local implementation gaps\n\n${factList(facts.gaps, 'Для topic немає actionable implementation gaps.')}${aliases}\n`
}

/**
 * Builds the optional architecture page without exposing private implementation.
 * @param {{graph: Record<string, unknown>, hiddenNames: Set<string>}} input render context
 * @returns {string} generated Markdown body
 */
function renderArchitecture({ graph, hiddenNames }) {
  const boundaries = graph.nodes
    .filter(node => node?.domainId === graph.domain.id && node.kind === 'integration' && node.visibility === 'external')
    .map(node => safeText(node.name, hiddenNames, 'External contract'))
    .toSorted(compareStrings)
  const lines =
    boundaries.length > 0
      ? boundaries.map(name => `- External boundary: ${name}`).join('\n')
      : '- Evidence-backed domain responsibility.'
  const architectureClaims = graph.claims
    .filter(claim => claim.layer === 'implemented' && ['responsibility', 'config', 'persistence', 'integration', 'state-change'].includes(claim.predicate))
    .toSorted((left, right) => compareStrings(left.id, right.id))
  if (architectureClaims.length > 0) {
    return `# Architecture: ${safeText(graph.domain.name, hiddenNames, 'Package domain')}\n\n## Implemented AS-IS\n\n${claimSections(architectureClaims, hiddenNames, 'Немає evidence-backed architecture claims.')}\n\n## Boundaries\n\n${lines}\n\n## Traceability\n\nManifest зберігає reverse evidence links до files, tests, configuration і contracts.\n`
  }
  return `# Architecture: ${safeText(graph.domain.name, hiddenNames, 'Package domain')}\n\n## Implemented AS-IS\n\nDomain architecture describes confirmed responsibilities and external boundaries without naming private implementation symbols.\n\n## Boundaries\n\n${lines}\n\n## Traceability\n\nThe manifest preserves reverse evidence links to files, tests, configuration and contracts.\n`
}

/* eslint-disable sonarjs/no-nested-template-literals -- String.raw is required for Mermaid quote escaping */
/**
 * Returns Mermaid only for a graph edge whose two visible endpoints are public.
 * @param {Record<string, unknown>} graph knowledge graph
 * @param {Set<string>} hiddenNames private names
 * @returns {string} Mermaid block or empty string
 */
function renderMermaid(graph, hiddenNames) {
  const nodes = new Map(
    graph.nodes
      .filter(node => node?.domainId === graph.domain.id && node.visibility !== 'private')
      .map(node => [node.id, node])
  )
  const edge = graph.edges.find(item => nodes.has(item?.fromId) && nodes.has(item?.toId))
  if (!edge) return ''
  const from = safeText(nodes.get(edge.fromId).name, hiddenNames, 'Source')
  const to = safeText(nodes.get(edge.toId).name, hiddenNames, 'Outcome')
  return `\n\n\`\`\`mermaid\nflowchart LR\n  source["${from.replaceAll('"', String.raw`\"`)}"] --> target["${to.replaceAll('"', String.raw`\"`)}"]\n\`\`\``
}
/* eslint-enable sonarjs/no-nested-template-literals */

/**
 * Builds the required package navigation page.
 * @param {{graph: Record<string, unknown>, topics: Array<Record<string, unknown>>, architecture: boolean, hiddenNames: Set<string>}} input render context
 * @returns {string} generated Markdown body
 */
function renderIndex({ graph, topics, architecture, hiddenNames }) {
  const groups = topicsByKind(topics)
  const sections = []
  if (architecture) sections.push('- [Architecture](explanation/architecture.md)')
  for (const kind of ['capability', 'process', 'contract']) {
    for (const topic of groups.get(kind) ?? []) {
      const path = `${PAGE_KIND_PATHS[kind].replace('docs/', '')}/${topicToken(topic.id)}.md`
      sections.push(`- [${safeText(topic.title, hiddenNames, PAGE_KIND_LABELS[kind])}](${path})`)
    }
  }
  const navigation =
    sections.length > 0 ? sections.join('\n') : '- Наразі graph не має evidence-backed dedicated topics.'
  return `# Package knowledge: ${safeText(graph.domain.name, hiddenNames, 'Package domain')}\n\n## Implemented AS-IS\n\nЦя навігація є deterministic projection одного documentation domain. Вона веде лише до meaningful pages і не розкриває private implementation symbols.\n\n## Views\n\n${navigation}\n\n## Traceability\n\n\`docs/.docgen/manifest.json\` містить stable topic identities, claims, evidence та reverse impact data.\n`
}

/**
 * Renders actionable (not satisfied) gap rows only.
 * @param {Record<string, unknown>} graph knowledge graph
 * @returns {string} generated Markdown body
 */
function renderGaps(graph) {
  const gaps = graph.gaps
    .filter(gap => gap.status !== 'satisfied')
    .toSorted((left, right) => compareStrings(left.id, right.id))
  const rows = gaps.map(gap => `- Status: ${gap.status}; explicit expectation requires review.`).join('\n')
  return `# Implementation gaps\n\n## Explicit expectation comparison\n\n${rows}\n\nOnly explicit expected claims participate in this view; absent expectations are not defects.\n`
}

/**
 * Renders candidate Markdown pages and a schema-compatible manifest.
 * @param {{graph: Record<string, unknown>, existingFiles?: Record<string, string>}} input graph plus committed page bytes
 * @returns {{ok: true, files: Record<string, string>} | {ok: false, diagnostics: Array<Record<string, unknown>>}} candidate map or blocking diagnostics
 */
export function renderKnowledgeArtifacts({ graph, existingFiles = {} }) {
  if (
    !graph ||
    typeof graph !== 'object' ||
    !graph.domain ||
    typeof graph.domain.id !== 'string' ||
    graph.domain.id === ''
  ) {
    return { ok: false, diagnostics: [diagnostic('invalid-render-graph', 'Graph має містити owning domain ID.')] }
  }
  if (!existingFiles || typeof existingFiles !== 'object' || Array.isArray(existingFiles)) {
    return {
      ok: false,
      diagnostics: [diagnostic('invalid-existing-files', 'existingFiles має бути path → Markdown map.')]
    }
  }
  const manifest = canonicalGraph(graph)
  const hiddenNames = privateNames(manifest)
  const topics = pageTopics(manifest)
  const architecture = needsArchitecture(manifest)
  const pages = [
    {
      path: 'docs/index.md',
      zoneId: 'package-index',
      content: renderIndex({ graph: manifest, topics, architecture, hiddenNames })
    }
  ]
  if (architecture) {
    pages.push({
      path: 'docs/explanation/architecture.md',
      zoneId: 'package-architecture',
      content: `${renderArchitecture({ graph: manifest, hiddenNames })}${renderMermaid(manifest, hiddenNames)}\n`
    })
  }
  for (const topic of topics) {
    const token = topicToken(topic.id)
    pages.push({
      path: `${PAGE_KIND_PATHS[topic.kind]}/${token}.md`,
      zoneId: `${topic.kind}-${token}`,
      content: renderTopic({ graph: manifest, topic, hiddenNames })
    })
  }
  if (manifest.gaps.some(gap => gap.status !== 'satisfied')) {
    pages.push({ path: 'docs/implementation-gaps.md', zoneId: 'implementation-gaps', content: renderGaps(manifest) })
  }

  const files = { [MANIFEST_PATH]: serializeKnowledgeGraph(manifest) }
  const diagnostics = []
  for (const page of pages) {
    const rendered = renderPage({ ...page, existing: existingFiles[page.path] })
    if (rendered.ok) {
      files[page.path] = rendered.markdown
    } else {
      diagnostics.push(...rendered.diagnostics)
    }
  }
  if (diagnostics.length > 0) {
    return {
      ok: false,
      diagnostics: diagnostics.toSorted((left, right) => compareStrings(left.path ?? '', right.path ?? ''))
    }
  }
  const humanMarkdown = Object.entries(files)
    .filter(([path]) => path.endsWith('.md'))
    .map(([, content]) => content)
    .join('\n')
  const leaked = [...hiddenNames].filter(name => humanMarkdown.includes(name))
  if (leaked.length > 0) {
    return {
      ok: false,
      diagnostics: leaked
        .toSorted(compareStrings)
        .map(name => diagnostic('private-symbol-leak', `Human Markdown містить private symbol name "${name}".`))
    }
  }
  return { ok: true, files }
}

File: crates/rules-docs/tests/render.rs
//! Дзеркальний набір рендерера — сценарії з `tests/render.test.mjs` плюс
//! ПОБАЙТОВА звірка всіх згенерованих файлів із живим JS
//! (`fixtures/js-render.json`).
//!
//! Побайтова звірка тут не примха: сторінки — це довгі шаблони українською,
//! і будь-яка розбіжність (пропущений абзац, інший fallback, зайвий перенос)
//! мовчки змінила б опубліковану документацію, а не впала б тестом.

use std::collections::BTreeMap;

use rules_docs::render::{render_knowledge_artifacts, RenderOutcome};
use rules_docs::zones::zone_hash;
use serde_json::{json, Value};

const FIXTURES: &str = include_str!("fixtures/js-render.json");
const DOMAIN: &str = "npm:@fixture/orders";
const PUBLIC_ID: &str = "code-unit:npm:@fixture/orders:js:src/orders.mjs#submitOrder";
const PRIVATE_ID: &str = "code-unit:npm:@fixture/orders:js:src/orders.mjs#persistOrder";
const OUTCOME_ID: &str = "outcome:created";
const CONTRACT_ID: &str = "contract:payments";

fn fixtures() -> Value {
    serde_json::from_str(FIXTURES).expect("фікстура — валідний JSON")
}

fn graph(with_gap: bool) -> Value {
    json!({
        "schemaVersion": 1,
        "domain": {"id": DOMAIN, "ecosystem": "npm", "name": "@fixture/orders",
                   "rootManifest": "package.json", "sourceFingerprint": "sha256:domain"},
        "nodes": [
            {"id": PUBLIC_ID, "kind": "code-unit", "name": "submitOrder", "visibility": "public",
             "domainId": DOMAIN, "attributes": {"sourcePath": "src/orders.mjs"}, "sourceFingerprint": "sha256:public"},
            {"id": PRIVATE_ID, "kind": "code-unit", "name": "persistOrder", "visibility": "private",
             "domainId": DOMAIN, "attributes": {"sourcePath": "src/persistence.mjs"}, "sourceFingerprint": "sha256:private"},
            {"id": OUTCOME_ID, "kind": "outcome", "name": "Order created", "visibility": "public",
             "domainId": DOMAIN, "attributes": {}, "sourceFingerprint": "sha256:outcome"},
            {"id": CONTRACT_ID, "kind": "integration", "name": "payments", "visibility": "external",
             "domainId": DOMAIN, "attributes": {}, "sourceFingerprint": "sha256:contract"}
        ],
        "edges": [{"id": "edge:public", "fromId": PUBLIC_ID, "toId": OUTCOME_ID, "kind": "produces",
                   "evidenceIds": ["e:public"]}],
        "claims": [
            {"id": "claim:implemented", "subjectId": PUBLIC_ID, "layer": "implemented", "predicate": "creates-order",
             "value": true, "evidenceIds": ["e:public"], "confidence": 1, "sourceFingerprint": "sha256:claim"},
            {"id": "claim:expected", "subjectId": PUBLIC_ID, "layer": "expected", "predicate": "creates-order",
             "value": true, "evidenceIds": ["e:public"], "confidence": 1, "sourceFingerprint": "sha256:expected"}
        ],
        "topics": [
            {"id": "process:orders", "kind": "process", "title": "submitOrder", "domainId": DOMAIN, "anchorIds": [PUBLIC_ID]},
            {"id": "contract:orders", "kind": "contract", "title": "payments", "domainId": DOMAIN, "anchorIds": [CONTRACT_ID]}
        ],
        "gaps": if with_gap {
            json!([{"id": "gap:expected", "status": "missing", "expectedClaimId": "claim:expected",
                    "implementedClaimIds": [], "evidenceIds": ["e:public"]}])
        } else {
            json!([])
        },
        "evidence": [{"id": "e:public", "kind": "code", "path": "src/orders.mjs", "symbolId": PUBLIC_ID,
                      "contentHash": "sha256:evidence"}]
    })
}

fn rendered(outcome: RenderOutcome) -> BTreeMap<String, String> {
    match outcome {
        RenderOutcome::Rendered(files) => files,
        RenderOutcome::Blocked(diagnostics) => panic!("несподівані блокери: {diagnostics:?}"),
    }
}

fn codes(outcome: RenderOutcome) -> Vec<String> {
    match outcome {
        RenderOutcome::Blocked(diagnostics) => {
            diagnostics.into_iter().map(|item| item.code).collect()
        }
        RenderOutcome::Rendered(files) => {
            panic!("очікувався блокер, отримано {:?}", files.keys())
        }
    }
}

fn empty_existing() -> BTreeMap<String, String> {
    BTreeMap::new()
}

#[test]
fn only_meaningful_views_plus_an_actionable_gaps_page_are_rendered() {
    let files = rendered(render_knowledge_artifacts(&graph(true), &empty_existing()));
    let paths: Vec<&String> = files.keys().collect();
    assert_eq!(
        paths,
        vec![
            "docs/.docgen/manifest.json",
            "docs/explanation/architecture.md",
            "docs/explanation/processes/dcfd264583ed8d3acfe0e103.md",
            "docs/implementation-gaps.md",
            "docs/index.md",
            "docs/reference/contracts/2e0b0c95a18292880dfd62a0.md"
        ]
    );
    assert!(files["docs/index.md"].contains("# Package knowledge: @fixture/orders"));
    assert!(files["docs/implementation-gaps.md"].contains("Status: missing"));
}

/// Порожніх дерев сторінок не буває: немає явної прогалини — немає й
/// сторінки прогалин.
#[test]
fn no_gaps_page_without_an_explicit_gap_and_the_output_is_deterministic() {
    let files = rendered(render_knowledge_artifacts(&graph(false), &empty_existing()));
    assert!(!files.contains_key("docs/implementation-gaps.md"));
    let expected: Vec<String> = fixtures()["noGapPaths"]
        .as_array()
        .expect("шляхи у фікстурі")
        .iter()
        .map(|item| item.as_str().unwrap_or_default().to_string())
        .collect();
    assert_eq!(files.keys().cloned().collect::<Vec<_>>(), expected);

    let again = rendered(render_knowledge_artifacts(&graph(false), &empty_existing()));
    assert_eq!(files, again, "двічі той самий вхід — той самий байт");
}

/// Приватне імʼя не має зʼявитись у людському Markdown НІ В ЯКОМУ вигляді —
/// ні як назва символу, ні як ID.
#[test]
fn private_names_never_leak_into_human_markdown() {
    let files = rendered(render_knowledge_artifacts(&graph(true), &empty_existing()));
    for (path, content) in &files {
        if !path.ends_with(".md") {
            continue;
        }
        assert!(
            !content.contains("persistOrder"),
            "приватне імʼя просочилось у {path}"
        );
    }
    assert!(
        files["docs/.docgen/manifest.json"].contains("persistOrder"),
        "у машинному manifest воно, навпаки, МУСИТЬ бути — це traceability"
    );
}

/// Авторський файл оновлюється лише в оголошеній AUTOGEN-зоні; MANUAL і
/// текст поза зонами лишаються байт-у-байт.
#[test]
fn authored_pages_keep_their_manual_zones_and_only_autogen_is_updated() {
    let authored = format!(
        "# Authored\n\n<!-- MANUAL:start id=\"note\" -->keep me<!-- MANUAL:end id=\"note\" -->\n<!-- AUTOGEN:start id=\"package-index\" hash=\"{}\" -->old<!-- AUTOGEN:end id=\"package-index\" -->\n",
        zone_hash("old")
    );
    let existing = BTreeMap::from([("docs/index.md".to_string(), authored.clone())]);
    assert_eq!(
        authored,
        fixtures()["authoredInput"].as_str().unwrap_or_default(),
        "вхідний авторський файл той самий, що в JS-фікстурі"
    );

    let files = rendered(render_knowledge_artifacts(&graph(true), &existing));
    let index = &files["docs/index.md"];
    assert!(index.contains("# Authored"), "заголовок автора лишився");
    assert!(index.contains("keep me"), "MANUAL лишився недоторканим");
    assert!(!index.contains(">old<"), "AUTOGEN оновлено");
    assert_eq!(
        index,
        fixtures()["authoredIndex"].as_str().unwrap_or_default(),
        "оновлений авторський файл збігається з JS побайтово"
    );
}

/// Авторська сторінка БЕЗ оголошеної AUTOGEN-зони — це помилка, а не привід
/// перезаписати її цілком.
#[test]
fn an_authored_page_without_an_autogen_target_fails_closed() {
    let existing = BTreeMap::from([(
        "docs/index.md".to_string(),
        "# Authored without zone\n".to_string(),
    )]);
    assert_eq!(
        codes(render_knowledge_artifacts(&graph(true), &existing)),
        vec!["autogen-zone-required".to_string()]
    );
}

#[test]
fn a_graph_without_a_domain_is_rejected() {
    assert_eq!(
        codes(render_knowledge_artifacts(
            &json!({"nodes": []}),
            &empty_existing()
        )),
        vec!["invalid-render-graph".to_string()]
    );
}

/// Головна звірка: КОЖЕН згенерований файл побайтово дорівнює JS-виходу.
#[test]
fn every_rendered_file_matches_the_js_renderer_byte_for_byte() {
    let expected = fixtures();
    let expected_files = expected["base"].as_object().expect("файли у фікстурі");
    let files = rendered(render_knowledge_artifacts(&graph(true), &empty_existing()));

    assert_eq!(
        files.len(),
        expected_files.len(),
        "набір файлів: {:?} проти {:?}",
        files.keys().collect::<Vec<_>>(),
        expected_files.keys().collect::<Vec<_>>()
    );
    for (path, content) in &files {
        let reference = expected_files
            .get(path)
            .and_then(Value::as_str)
            .unwrap_or_else(|| panic!("у JS-фікстурі немає файла {path}"));
        assert_eq!(content, reference, "вміст {path} розійшовся з JS");
    }
}

#[test]
fn the_blocking_diagnostics_match_the_js_renderer() {
    let existing = BTreeMap::from([(
        "docs/index.md".to_string(),
        "# Authored without zone\n".to_string(),
    )]);
    let reported = match render_knowledge_artifacts(&graph(true), &existing) {
        RenderOutcome::Blocked(diagnostics) => diagnostics
            .into_iter()
            .map(|item| json!({"code": item.code, "detail": item.detail, "path": item.path}))
            .collect::<Vec<_>>(),
        RenderOutcome::Rendered(_) => panic!("очікувався блокер"),
    };
    assert_eq!(Value::Array(reported), fixtures()["missingZone"]);
}
